### PR TITLE
ROCKernels: Add occupancy API

### PR DESCRIPTION
**Pull request by pxl-th**
Relies on https://github.com/JuliaGPU/AMDGPU.jl/pull/326 for occupancy API.

### CODE REVIEW

**Comment by simeonschaub** [comment]
Would it make sense to move most of the logic here to KA base, so this code is shared between CUDA and AMDGPU? It looks to me like they're doing basically the same thing, so that might make it a little easier to maintain? Doesn't need to block this PR of course

**Comment by pxl-th** [comment]
> Would it make sense to move most of the logic here to KA base, so this code is shared between CUDA and AMDGPU? It looks to me like they're doing basically the same thing, so that might make it a little easier to maintain? Doesn't need to block this PR of course

I guess we'd need to define some abstractions for compiling kernels, using occupancy, etc.

**Comment by pxl-th** [comment]
Since https://github.com/JuliaGPU/AMDGPU.jl/pull/326 was merged and if there are no objections, it'd be good to have it merged.
Or do we need to tag release for AMDGPU first?

@jpsamaroo

**Comment by vchuravy** [comment]
We need a release of AMDGPU first

**Comment by pxl-th** [comment]
> We need a release of AMDGPU first

Done

**Comment by vchuravy** [comment]
Can you split the PRs? The occupancy work is fine to go on as is, but the pooling one requires a bit more thought.

**Review by vchuravy — APPROVED** [review]

**Review by jpsamaroo — APPROVED** [review]

File: lib/ROCKernels/src/ROCKernels.jl
Comment (pxl-th) [reply lib/ROCKernels/src/ROCKernels.jl:227]: I guess I wanted to proceed with the kernel we get from the first call to `@roc`.
As it does slightly less allocations, but not that significant:
- "manual" launch: `0.005425 seconds (255 allocations: 9.031 KiB)`
- calling `@roc` for launch: `0.005892 seconds (296 allocations: 10.422 KiB)`

Comment (jpsamaroo) [reply lib/ROCKernels/src/ROCKernels.jl:227]: I guess that's fine to keep as-is; maybe we can consider adding support to `@roc` to accept a `HostKernel` object? But that can be done later.

Comment (vchuravy) [reply lib/ROCKernels/src/ROCKernels.jl:227]: Please revert to `@roc` until this package moves to AMDGPU.jl proper.

Comment (pxl-th) [reply lib/ROCKernels/src/ROCKernels.jl]: > Please revert to `@roc` until this package moves to AMDGPU.jl proper.

Done

**Review by vchuravy — APPROVED** [review]

**Comment by vchuravy** [comment]
Can you rebase on master?

**Comment by pxl-th** [comment]
> Can you rebase on master?

Done.

**Comment by vchuravy** [comment]
Note for future merges, we prefer squash merge to keep the history clean or rebase and squash manually before merge. Except when the commits are logically separate.

**Comment by pxl-th** [comment]
> Note for future merges, we prefer squash merge to keep the history clean or rebase and squash manually before merge. Except when the commits are logically separate.

Got it!

**Pull request merged**
